### PR TITLE
acme.jose: (Typed)JSONObjectWithFields, Field, JWA.

### PR DESCRIPTION
**Pull request by kuba**
It's probably best to start with `acme.jose.interfaces` and `acme.jose.json_until`: both define well documented, more structured and unified framework for JSON de(serialization). `acme.messages`/`acme.challenges` have been changed accordingly and are now more compact and hopefully more readable.

Apart from the above refactor there is a whole new support for all JWA signature algorithms, but ES* family for elliptic curves (which library should be used for that?). JSON schema from `schemata` directory is yet to be updated.

I will submit JWS PR shortly.

### CODE REVIEW

File: letsencrypt/acme/jose/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+              return Foo()
+
+      class Bar(JSONDeSerializable):
+          def to_json(self):
+              return [Foo(), Foo()]
+
+          @classmethod
+          def from_json(cls, jobj):
+              return Bar()
+
```
Comment (jdkasten) [inline letsencrypt/acme/jose/interfaces.py]: Thanks for all of the details!

File: letsencrypt/acme/challenges_test.py [hunk 2]
```diff
@@ -236,37 +234,37 @@ def test_json_without_optionals(self):
...
     def setUp(self):
-        jwk = other.JWK(key=KEY.publickey())
-        issuers = [
+        jwk = jose.JWKRSA(key=KEY.publickey())
+        issuers = (
             'C=US, O=SuperT LLC, CN=SuperTrustworthy Public CA',
             'O=LessTrustworthy CA Inc, CN=LessTrustworthy But StillSecure',
-        ]
-        cert_fingerprints = [
+        )
```
Comment (jdkasten) [inline letsencrypt/acme/challenges_test.py]: I see how the tuples work, but it isn't what I would expect... The lists seem more appropriate.  My guess is that the lists caused problems with json serialization/deserialization???

File: letsencrypt/acme/jose/jwa.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    def __init__(self, name, digestmod):
+        super(_JWAHS, self).__init__(name)
+        self.digestmod = digestmod
+
+    def sign(self, key, msg):
+        return HMAC.new(key, msg, self.digestmod).digest()
+
+    def verify(self, key, msg, sig):
+        # TODO: use constant compare to mitigate timing attack?
```
Comment (jdkasten) [inline letsencrypt/acme/jose/jwa.py]: This is more important for server side code... however, I am afraid that this library code might be adopted somewhere, given that a good python implementation doesn't already exist.  I would personally make it a warning.
http://www.levigross.com/2014/02/07/constant-time-comparison-functions-in-python-haskell-clojure-java-etc/

Comment (jdkasten) [inline letsencrypt/acme/jose/jwa.py]: Also, thanks for recognizing this! :+1:

File: letsencrypt/acme/jose/jwk.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    @classmethod
+    def fields_from_json(cls, jobj):
+        return cls(key=jobj['k'])
+
+
+@JWK.register
+class JWKRSA(JWK):
+    """RSA JWK."""
+    typ = 'RSA'
+    __slots__ = ('key',)
```
Comment (jdkasten) [inline letsencrypt/acme/jose/jwk.py]: These don't use the jose.fields method?  Too much complication?

**Comment by jdkasten** [comment]
I was interrupted quite a few times as i made my way through the code.  I feel i have a decent understanding now but will have to make another pass to make real comments. It might be easier to finish this up in person tomorrow... perhaps if you are free we can set up a call and knock the rest of this out.

File: letsencrypt/acme/challenges_test.py (hunk 2 above)
Comment (kuba) [inline letsencrypt/acme/challenges_test.py]: My aim was to make all protocol objects (messages, challenges, and anything else down the stack) hashable (`__hash__`), which enables to use those objects e.g. as keys in dictionaries (useful especially for server, which has to persist and match up session ACME messages/challenges). This in turn requires the objects to be immutable. That's why I'm using `tuple` over `list`, `frozendict` over `dict`, `frozenset` over `set`...

File: letsencrypt/acme/jose/jwk.py (hunk 4 above)
Comment (kuba) [inline letsencrypt/acme/jose/jwk.py]: `jose.Field` is used for establishing 1-1 relation between JSON object field and internal representation "slot" (attribute): for `JWK` that would be `n = jose.Field('n', ...); e = jose.Field('e', ...)`. It seems more convenient to unpack `n` and `e` into `Crypto.PublicKey.RSA` already at the deserialization step (similarily for the serialization): that is why I override `fields_from_json` and `fields_to_json`. Does this make sense?

**Comment by kuba** [comment]
Yup, I'll be ready for the Hangout tonight, just send me a link :) Thank you so much for reviewing!

File: letsencrypt/acme/jose/jwa.py (hunk 3 above)
Comment (kuba) [inline letsencrypt/acme/jose/jwa.py]: No problem, made it a warning in the latest commit.

File: letsencrypt/acme/jose/json_util.py [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+    :ivar bool omitempty: If ``True`` and the field value is empty, then
+        it will not be included in the serialized JSON object, and
+        ``default`` will be used for deserialization. Otherwise, if ``False``,
+        field is considered as required, value will always be included in the
+        serialized JSON objected, and it must also be present when
+        deserializing.
+
+    """
+    __slots__ = ('json_name', 'default', 'omitempty',
+                 'fdec', 'fenc', 'fdec2', 'fenc2')
```
Comment (jdkasten) [inline letsencrypt/acme/jose/json_util.py]: I believe fdec2 and fenc2 are remnants of encoder2/decoder2

File: letsencrypt/acme/messages.py [hunk 6]
```diff
@@ -96,86 +60,55 @@ class Challenge(Message):
     :ivar list challenges: List of
         :class:`~letsencrypt.acme.challenges.Challenge` objects.
 
+    .. todo::
+        1. can challenges contain two challenges of the same type?
+        2. can challenges contain duplicates?
+        3. check "combinations" indices are in valid range
+        4. turn "combinations" elements into sets?
+        5. turn "combinations" into set?
+
```
Comment (jdkasten) [inline letsencrypt/acme/messages.py]: 1. Yes
2. No, I don't believe there is a need for duplicate challenges, though it is supported technically in the spec.
4. 5. Both would be fine.

File: tox.ini [hunk 7]
```diff
@@ -12,6 +12,8 @@ commands =
 
 setenv =
     PYTHONPATH = {toxinidir}
+    PYTHONHASHSEED = 0
+# https://testrun.org/tox/latest/example/basic.html#special-handling-of-pythonhas
```
Comment (jdkasten) [inline tox.ini]: Thanks for the link

File: letsencrypt/acme/jose/json_util.py (hunk 5 above)
Comment (kuba) [inline letsencrypt/acme/jose/json_util.py]: thanks, fixed!

File: letsencrypt/acme/messages.py (hunk 6 above)
Comment (kuba) [inline letsencrypt/acme/messages.py]: Great, thanks for explanation. Will send it in a follow-up PR :)

**Comment by jdkasten** [comment]
Great! Thanks, Kuba!

**Pull request merged**